HLTrigger/HLTcore/interface/TriggerExpressionParser.h

```diff
@@ -23,11 +23,22 @@ namespace triggerExpression {
   class Parser : public qi::grammar<Iterator, Evaluator *(), ascii::space_type> {
   public:
     Parser() : Parser::base_type(expression) {
+      auto delimiter = qi::copy(qi::eoi | !qi::char_("a-zA-Z0-9_*?"));
+
+      token_true = qi::lexeme[qi::lit("TRUE") >> delimiter];
+      token_false = qi::lexeme[qi::lit("FALSE") >> delimiter];
+
+      operand_not = qi::lexeme[qi::lit("NOT") >> delimiter];
+      operand_and = qi::lexeme[qi::lit("AND") >> delimiter];
+      operand_or = qi::lexeme[qi::lit("OR") >> delimiter];
+
       token_l1algo %= qi::raw[qi::lexeme["L1_" >> +(qi::char_("a-zA-Z0-9_*?"))]];
       token_path %= qi::raw[qi::lexeme[+(qi::char_("a-zA-Z0-9_*?"))]];
```

Comment:
`AND`, `OR` and `NOT` can in principle be interpreted as names of HLT paths, for example the parser gives
```
THIS OR OR AND NOT-> ((Uninitialised_Path_Expression OR Uninitialised_Path_Expression) AND Uninitialised_Path_Expression)
```
I don't have a strong opinion on whether this should be allowed, but I'm leaning towards leaving it as valid.
 - One could in principle have `cms.Path`s named `AND/OR/NOT`, and the parser is currently able to support that.
 - If there is no Path named `AND/OR/NOT` and `throw=True`, `TriggerResultsFilter` will anyway return an exception for cases like `THIS OR OR AND NOT` (and if `throw=False`, one is intentionally disabling checks, in a sense).

Am I missing anything?

PS. One way to make it invalid is the following (I checked this with the current unit tests)
```diff
diff --git a/HLTrigger/HLTcore/interface/TriggerExpressionParser.h b/HLTrigger/HLTcore/interface/TriggerExpressionParser.h
index a47505c82e0..5c107d16c09 100644
--- a/HLTrigger/HLTcore/interface/TriggerExpressionParser.h
+++ b/HLTrigger/HLTcore/interface/TriggerExpressionParser.h
@@ -33,12 +33,12 @@ namespace triggerExpression {
       operand_or = qi::lexeme[qi::lit("OR") >> delimiter];
 
       token_l1algo %= qi::raw[qi::lexeme["L1_" >> +(qi::char_("a-zA-Z0-9_*?"))]];
-      token_path %= qi::raw[qi::lexeme[+(qi::char_("a-zA-Z0-9_*?"))]];
+      token_path %= qi::raw[qi::lexeme[+(qi::char_("a-zA-Z0-9_*?"))] - operand_not - operand_and - operand_or];
 
       token = (token_true[qi::_val = new_<Constant>(true)] |         // TRUE
                token_false[qi::_val = new_<Constant>(false)] |       // FALSE
                token_l1algo[qi::_val = new_<L1uGTReader>(qi::_1)] |  // L1_*
-               token_path[qi::_val = new_<PathReader>(qi::_1)]);     // *
+               token_path[qi::_val = new_<PathReader>(qi::_1)]);     // * (except "NOT", "AND" and "OR")
 
       parenthesis %= ('(' >> expression >> ')');
```

Reply:
I would prefer to make NOT, AND, OR invalid as Path names.
If they have to be blacklisted explicitly I have the feeling that something is wrong with the grammar, but I don't have a better suggestion at the moment.

Reply:
Thanks for the feedback. I added 608afea with the solution I currently know. I will try to find a better one, but I would consider 608afea acceptable, since all the improved unit tests pass.